packages/jest-jasmine2/src/index.js

```diff
@@ -39,6 +39,28 @@ function jasmine2(config, environment, moduleLoader, testPath) {
 
     const requireJasmine = environment.global.jasmineRequire;
     jasmine = requireJasmine.core(requireJasmine);
+
+    const jasmineBuildExpectationResult = jasmine.buildExpectationResult;
+
+    // @see https://github.com/facebook/jest/issues/429
+    //
+    // Because we build error messages after the test is complete we can run
+    // into reporting issues.
+    // @example
+    //    let a = [1];
+    //    expect(a).toEqual([1, 2]);
+    //    a.push(2); // mutate the original value after it fails the matcher
+    // will print `expected [2] to equal [2]`
+    //
+    // The workaround is to clone the values at the time of the failure. That
+    // way we hold the value that won't be mutated during the test execution.
+    jasmine.buildExpectationResult = function(options) {
+      options.expected = JSON.parse(JSON.stringify(options.expected));
+      options.actual = JSON.parse(JSON.stringify(options.actual));
```

Comment:
I'm worried that accessing .message will call the getters for our custom matchers and slow things down. Probably makes sense to enforce checking passed first and only if it's not passed, check the message - what do you think Cristian?





On Wed, Apr 13, 2016 at 3:28 AM -0700, "Cristian Carlesso" <notifications@github.com<mailto:notifications@github.com>> wrote:


In packages/jest-jasmine2/src/index.js<https://github.com/facebook/jest/pull/889#discussion_r59526074>:

> +
> +    // @see https://github.com/facebook/jest/issues/429
> +    //
> +    // Because we build error messages after the test is complete we can run
> +    // into reporting issues.
> +    // @example
> +    //    let a = [1];
> +    //    expect(a).toEqual([1, 2]);
> +    //    a.push(2); // mutate the original value after it fails the matcher
> +    // will print `expected [2] to equal [2]`
> +    //
> +    // The workaround is to clone the values at the time of the failure. That
> +    // way we hold the value that won't be mutated during the test execution.
> +    jasmine.buildExpectationResult = function(options) {
> +      options.expected = JSON.parse(JSON.stringify(options.expected));
> +      options.actual = JSON.parse(JSON.stringify(options.actual));


I used reporter.getFormatter().prettyPrint in the matchers for doing something similar,
also looks like expected and actual should only be set when options.passed is false https://github.com/jasmine/jasmine/blob/master/src/core/ExpectationResult.js#L24
it would be probably better to patch the results only in that case, also is worth checking if options.message exists as that should indicate a custom matcher (to avoid double patching it)

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub<https://github.com/facebook/jest/pull/889/files/e3aa69cd8382167a5a9ca47b6a5921b19cc14079#r59526074>